File: Dynamic Programming/Book_shop.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;
#define ll long long int
const int MOD = 1000000007;

int main(){
    ll n,x;
    cin >> n >> x;
    vector<int> prices(n);
    vector<int> pages(n);
    for(ll i=0;i<n;i++){
        cin >> prices[i];
    }
    for(ll i=0;i<n;i++){
        cin >> pages[i];
    }
    ll dp[n+1][x+1];
    for(ll i=0;i<n+1;i++){
        for(ll j=0;j<x+1;j++){
            if (i == 0 || j == 0)
                dp[i][j] = 0;
            else if (j >= prices[i-1])
                dp[i][j] = max(dp[i-1][j],pages[i-1] + dp[i-1][j-prices[i-1]]);
            else
                dp[i][j] = dp[i-1][j];
            
        }
    }
    cout << dp[n][x] <<'\n';
    return 0;
}

File: Dynamic Programming/Book_shop.py
def helper(n,x,prices,pages):
    dp = [[0 for i in range(x+1)] for i in range(n+1)]
    for i in range(n+1):
        for j in range(x+1):
            if i == 0 or j == 0:
                dp[i][j] = 0
            elif prices[i-1] <= j:
                dp[i][j] = max(pages[i-1] + dp[i-1][j-prices[i-1]],dp[i-1][j])
            else:
                dp[i][j] = dp[i-1][j]
    return dp[n][x]

n,x = map(int,raw_input().split())
prices = list(map(int,raw_input().split()))
pages = list(map(int,raw_input().split()))
print(helper(n,x,prices,pages))